src/spi.rs

```diff
@@ -308,29 +308,142 @@ impl<'d> SpiBusDriver<'d> {
     }
 }
 
-impl<'d> embedded_hal::spi::ErrorType for SpiBusDriver<'d> {
+impl<'d> embedded_hal::spi::ErrorType for SpiSharedBusDriver<'d> {
     type Error = SpiError;
 }
 
-impl<'d> SpiBusFlush for SpiBusDriver<'d> {
+impl<'d> SpiBusFlush for SpiSharedBusDriver<'d> {
+    fn flush(&mut self) -> Result<(), Self::Error> {
+        SpiSharedBusDriver::flush(self).map_err(to_spi_err)
+    }
+}
+
+impl<'d> SpiBusRead for SpiSharedBusDriver<'d> {
+    fn read(&mut self, words: &mut [u8]) -> Result<(), Self::Error> {
+        SpiSharedBusDriver::read(self, words).map_err(to_spi_err)
+    }
+}
+
+impl<'d> SpiBusWrite for SpiSharedBusDriver<'d> {
+    fn write(&mut self, words: &[u8]) -> Result<(), Self::Error> {
+        SpiSharedBusDriver::write(self, words).map_err(to_spi_err)
+    }
+}
+
+impl<'d> SpiBus for SpiSharedBusDriver<'d> {
+    fn transfer(&mut self, read: &mut [u8], write: &[u8]) -> Result<(), Self::Error> {
+        SpiSharedBusDriver::transfer(self, read, write).map_err(to_spi_err)
+    }
+
+    fn transfer_in_place(&mut self, words: &mut [u8]) -> Result<(), Self::Error> {
+        SpiSharedBusDriver::transfer_in_place(self, words).map_err(to_spi_err)
+    }
+}
+
+struct Device(spi_device_handle_t);
+
+impl Drop for Device {
+    fn drop(&mut self) {
+        esp!(unsafe { spi_bus_remove_device(self.0) }).unwrap();
+    }
+}
+
+pub struct SpiBusDriver<T> {
```

Comment:
Instead of implementing yet another structure, which is (again!) containing `T` which we are trying to erase everywhere, how about the following:
* Just one struct - `SpiBusDriver`. Convers all three cases:
  * Passed to the user in his/her `transaction` closure
  * Returned from `into_bus()` - see below
  * Returned from `bus()` - see below

Two new methods:
* `into_bus(self)` - transforms an `SpiDriver` into an `SpiBusDriver` by consuming `self`. The `'d` lifetime remains the same
* `bus<'a>(&'a mut self) -> Result<SpiBusDriver<'a>, EspError>` - takes a `&mut` ref to `SpiDriver` and produces an `SpiBusDriver`. Can be used when the original `SpiDriver` needs to be recovered after the "exclusive" bus operation.

Possible? Should be?

Reply:
The struct consolidation makes sense and I quite like it, however, the issue with consolidation is ownership.
`SpiBusDriver` owns a device and `SpiSharedBusDriver` (pending rename) borrows a device.
What would the implementation of `Drop` be?

I find `into_bus` and `bus` a bit limiting, i.e. I kinda have a `Borrow(Mut)` obsession aha.
Your example `bus<'a>(&'a mut self) -> Result<SpiBusDriver<'a>, EspError>` still has generics in it.
Does `SpiBusDriver<'a>` look better than `SpiBusDriver<T>`? The latter is more flexible.
Thinking about it more though, I suppose the flexibility isn't really needed for a bus. I just don't want to regret making this inflexible.

Reply:
And yeah, by that logic, in the majority of the cases, your `SpiBusDriver<T>` becomes `SpiBusDriver<SpiDriver<'static>>`, but that's a bit mouthful. 

Now that I think of it, I probably don't have that much of an issue with the `T` generic, as much as I do with having yet another structure (`SpiSharedBusDriver`) which is not even having a good name IMO, as this is really not shared. :)